lib/navigation/preprocess_imu.cpp

```diff
@@ -44,16 +46,53 @@ namespace hyped::navigation
       {
         magnitude += std::pow(imu_data.at(i).at(j), 2);
       }
-      // TODO change to sqrt function
-      clean_imu_data.at(i) = std::pow(magnitude, 0.5);
+      clean_accelerometer_data.at(i) = std::sqrt(magnitude);
     }
-    const uint8_t num_reliable_imus = std::accumulate(are_imus_reliable_.begin(), are_imus_reliable_.end(), 0);
-    if (num_reliable_imus == 4)
-    {
-      std::sort(imu_data.start(), imu_data.end());
-      const core::Float q1 = (imu_data.at(0) + imu_data.at(1)) / 2.0;
-      const core::Float median = (imu_data.at(1) + imu_data.at(2)) / 2.0;
-      const core::Float q3 = (imu_data.at(2) + imu_data.at(3)) / 2.0;
+    const uint8_t num_reliable_accelerometers = std::accumulate(are_imus_reliable_.begin(), are_imus_reliable_.end(), 0);
+    if (num_reliable_accelerometers == 4)
+    { 
+      core::ImuData clean_accelerometer_data_copy = std::copy(clean_accelerometer_data);
+      std::sort(clean_accelerometer_data_copy.begin(), clean_accelerometer_data_copy.end());
+      const core::Float q1 = (clean_accelerometer_data_copy.at(0) + clean_accelerometer_data_copy.at(1)) / 2.0;
+      const core::Float median = (clean_accelerometer_data_copy.at(1) + clean_accelerometer_data_copy.at(2)) / 2.0;
+      const core::Float q3 = (clean_accelerometer_data_copy.at(2) + clean_accelerometer_data_copy.at(3)) / 2.0;
+      const core::Float iqr = q3 - q1;
+
+      for (int i = 0; i < core::kNumImus; ++i) {
+        if (clean_accelerometer_data.at(i) > q3 + 1.5*iqr || clean_accelerometer_data.at(i) < q1 - 1.5*iqr) {
+            clean_accelerometer_data.at(i) = median;
+            num_outliers_per_imu_.at(i) += 1;
+        } else {
+          num_outliers_per_imu_.at(i) = 0;
+        }
+      }
+    } else if (num_reliable_accelerometers == 3) {
+      core::ImuData clean_accelerometer_data_copy = std::copy(clean_accelerometer_data);
+      for (int i = 0; i < core::kNumImus; ++i) {
+        if (are_imus_reliable_.at(i) == false) {
+          clean_accelerometer_data_copy.at(i) = -1;
```

Comment:
I'm not  fan of this method. Think of a way to do it without needing to do this.